tmux/tmux
# Add cursor style option

~~🚧 The changes in this PR can cause tmux spin, i.e. use ~95%, this is a known issue and is being worked on. Until this is resolved this PR should not be merged.~~ ✅ resolved.

~~⚠️ This PRs merge base should be [afh:afh-cursor-color](https://github.com/afh/tmux/tree/afh-cursor-colour); [here are the relevant changes between afh:afh-cursor-color and this PRs branch](https://github.com/afh/tmux/compare/afh-cursor-colour...afh:afh-cursor-style?expand=1)~~ **Update:** [afh:afh-cursor-color](https://github.com/afh/tmux/tree/afh-cursor-colour) has been merged with #2959 and using master as this PRs merge is now correct.

To test:
* Build and run tmux: `./tmux -vv -f/dev/null -Ltest`
* Set the different styles for the cursor: `./tmux set cursor-style $style` where `$style` is on of:`default`, `block-blinking`, `block`, `underline-blinking`, `underline`, `bar-blinking`, `bar`
* Use SM/RM ?12 to set cursor (not) blinking `echo -ne "\033[?12h\a"` / `echo -ne "\033[?12l\a"`
* Use DECSCUSR to change cursor style `echo -en "\033[4 q\a"`

ℹ️ If the changes in this PR are applied the **Options for default cursor style.** item can be removed from the [Contribution — Small things](https://github.com/tmux/tmux/wiki/Contributing#small-things) wiki page

<details id=test_cursor_style><summary><code>test_cursor_style.sh</code></summary>

```
#!/bin/ksh

echo "Cursor should be default. Enter to continue"; read

./tmux set cursor-style bar-blinking
echo "tmux cursor-style set to bar-blinking. Enter to continue"; read

./tmux set -u cursor-style
echo "tmux cursor-style unset. Cursor should be default. Enter to continue"; read

echo -en "\033[4 q\a"
echo "DECSCUSR 4. Cursor should be underline. Enter to continue"; read

echo -ne "\033[?12h\a"
echo "SM 12. Cursor should be blinking. Enter to continue"; read

echo -en "\033[0 q\a"
echo "DECSCUSR 0. Cursor should be default blinking. Enter to continue"; read

./tmux set cursor-style bar
echo "tmux cursor-style set to bar. Cursor should still be blinking due to SM 12. Enter to continue"; read

echo -ne "\033[?12l\a"
echo "RM 12.Cursor should not be blinking. Enter to continue"; read

./tmux set cursor-style block-blinking
echo "tmux cursor-style set to block-blinking. Cursor should not be blinking due to RM 12. Enter to continue"; read

./tmux set -u cursor-style
echo "tmux cursor-style unset. Cursor should be default. Enter to continue"; read
```

</details>
